### Advanced Big Data and Data Mining

### Assignment 2: Advanced Data Mining Techniques: Clustering, Classification, and Association Rule Mining

**Import Libraries**

Importing necessary libraries for data processing, computation, plotting and machine learning has been successfully done. The environment was ready to process data, discover patterns, classify and present the results in the assignment.

In [19]:
# importing the required libraries
import pandas as pd
import matplotlib.pyplot as plt
import warnings
import pandas as pd
import numpy as np
from math import log2
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier, export_text
from sklearn.naive_bayes import CategoricalNB

from sklearn.tree import DecisionTreeClassifier
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, fpgrowth, association_rules

**Create Transaction Dataset**

The data from the transaction was put into a tabular format where each transaction was a row and the presence or absence of each item was represented as a column. Prepared data was presented to ensure that the prepared transaction records have been converted correctly for further analysis.

In [3]:
# Creating transaction dataset
transactions = [
    ['A','B','C','D'],
    ['A','C','E'],
    ['B','C','D','E'],
    ['A','B','D','E'],
    ['A','B','C']
]

# Converting transactions into encoder format
te = TransactionEncoder()

# Transforming transactions into boolean values
te_array = te.fit(transactions).transform(transactions)

# Creating dataframe from transformed data
df = pd.DataFrame(te_array, columns=te.columns_)

# Displaying transaction dataframe
print(df)

       A      B      C      D      E
0   True   True   True   True  False
1   True  False   True  False   True
2  False   True   True   True   True
3   True   True  False   True   True
4   True   True   True  False  False


**Apriori Algorithm**

Transaction data was used to identify the frequent itemsets which met the minimum support requirement. The pairs of items and the support percentage associated with each pair were presented for further association analysis.

In [4]:
# Finding frequent itemsets using Apriori algorithm
frequent_itemsets_ap = apriori(
    df,
    min_support=0.5,
    use_colnames=True
)

# Displaying frequent itemsets
print("Frequent Itemsets using Apriori")
print(frequent_itemsets_ap)

Frequent Itemsets using Apriori
   support           itemsets
0      0.8     frozenset({A})
1      0.8     frozenset({B})
2      0.8     frozenset({C})
3      0.6     frozenset({D})
4      0.6     frozenset({E})
5      0.6  frozenset({A, B})
6      0.6  frozenset({C, A})
7      0.6  frozenset({C, B})
8      0.6  frozenset({D, B})


**FP-Growth Algorithm**

The minimum support was set and the frequent itemsets were found using the transaction data. The item combinations and their supports were then shown in comparison to the previous method.

In [5]:
# Finding frequent itemsets using FP Growth algorithm
frequent_itemsets_fp = fpgrowth(
    df,
    min_support=0.5,
    use_colnames=True
)

# Displaying frequent itemsets
print("Frequent Itemsets using FP-Growth")
print(frequent_itemsets_fp)

Frequent Itemsets using FP-Growth
   support           itemsets
0      0.8     frozenset({C})
1      0.8     frozenset({A})
2      0.8     frozenset({B})
3      0.6     frozenset({D})
4      0.6     frozenset({E})
5      0.6  frozenset({C, A})
6      0.6  frozenset({A, B})
7      0.6  frozenset({C, B})
8      0.6  frozenset({D, B})


**Generate Association Rules**

Strong association rules where they met the minimum confidence were extracted from the frequent itemsets. The relationships along with the values regarding confidence, support and strength were shown for interpretation.

In [6]:
# Generating association rules using confidence
rules = association_rules(
    frequent_itemsets_ap,
    metric="confidence",
    min_threshold=0.7
)

# Displaying strong association rules
print("Strong Association Rules")

print(rules[
    ['antecedents',
     'consequents',
     'support',
     'confidence',
     'lift']
])

Strong Association Rules
      antecedents     consequents  support  confidence    lift
0  frozenset({A})  frozenset({B})      0.6        0.75  0.9375
1  frozenset({B})  frozenset({A})      0.6        0.75  0.9375
2  frozenset({C})  frozenset({A})      0.6        0.75  0.9375
3  frozenset({A})  frozenset({C})      0.6        0.75  0.9375
4  frozenset({C})  frozenset({B})      0.6        0.75  0.9375
5  frozenset({B})  frozenset({C})      0.6        0.75  0.9375
6  frozenset({D})  frozenset({B})      0.6        1.00  1.2500
7  frozenset({B})  frozenset({D})      0.6        0.75  1.2500


**Analysis and Interpretation**

**Insights from the Generated Rules**

The generated association rules indicate that items A, B, and C are frequently purchased together, showing a strong relationship among these products. The rule D ⇒ B has the highest confidence (100%), meaning every customer who purchased D also purchased B, making it the strongest rule. These patterns reveal customer buying behavior and identify products that are commonly bought together.

**Application in a Retail Environment**

Retailers can use these rules to create bundle offers, cross-selling, and personalized product recommendations to increase sales. Products with strong associations can be placed near each other on store shelves, and inventory managers can ensure sufficient stock of frequently co-purchased items to reduce stock-outs and improve customer satisfaction.

### 2: Association Rule Mining and Correlation Analysis

Association and correlation analysis threshold values and transaction count were determined. The necessary values were obtained for calculating the support, confidence, independence, etc. and other measures of relationship.

In [1]:
# Importing math library
import math

# Storing total number of transactions
total_transactions = 5000

# Storing fiction transaction count
fiction = 2500

# Storing non fiction transaction count
not_fiction = 2500

# Storing mystery transaction count
mystery = 2400

# Storing non mystery transaction count
not_mystery = 2600

# Storing fiction and mystery transaction count
fiction_and_mystery = 1800

# Storing fiction and non mystery transaction count
fiction_and_not_mystery = 700

# Storing non fiction and mystery transaction count
not_fiction_and_mystery = 600

# Storing non fiction and non mystery transaction count
not_fiction_and_not_mystery = 1900

# Setting minimum support threshold
min_support = 0.30

# Setting minimum confidence threshold
min_confidence = 0.60

**Association Rule: Fiction ⇒ Mystery**

Association rule support and confidence values were calculated and compared to threshold values provided. Based on the result, it was decided whether the rule met the necessary conditions to be a strong association rule or not.

In [6]:
# Displaying heading for strong association rule
print("(a) Strong Association Rule")

# Calculating support value
support = fiction_and_mystery / total_transactions

# Calculating confidence value
confidence = fiction_and_mystery / fiction

# Displaying support calculation
print(f"Support = {fiction_and_mystery}/{total_transactions}")
print(f"Support = {support:.4f} ({support*100:.2f}%)")

# Displaying blank line
print()

# Displaying confidence calculation
print(f"Confidence = {fiction_and_mystery}/{fiction}")
print(f"Confidence = {confidence:.4f} ({confidence*100:.2f}%)")

# Displaying blank line
print()

# Displaying minimum support threshold
print("Minimum Support Threshold :", min_support*100,"%")

# Displaying minimum confidence threshold
print("Minimum Confidence Threshold :", min_confidence*100,"%")

# Displaying blank line
print()

# Checking rule strength
if support >= min_support and confidence >= min_confidence:
    print("Result : Fiction ⇒ Mystery is a STRONG association rule.")
else:
    print("Result : Fiction ⇒ Mystery is NOT a strong association rule.")

(a) Strong Association Rule
Support = 1800/5000
Support = 0.3600 (36.00%)

Confidence = 1800/2500
Confidence = 0.7200 (72.00%)

Minimum Support Threshold : 30.0 %
Minimum Confidence Threshold : 60.0 %

Result : Fiction ⇒ Mystery is a STRONG association rule.


**Independence and Correlation Analysis**

The probabilities of customers buying each book type individually and together were determined to see if the two book types were independent. The comparison revealed the nature of the relationship and whether it was a positive or negative correlation between the purchases.

In [5]:
# Displaying heading for independence and correlation analysis
print("(b) Independence and Correlation Analysis")

# Calculating fiction probability
P_fiction = fiction / total_transactions

# Calculating mystery probability
P_mystery = mystery / total_transactions

# Calculating joint probability
P_both = fiction_and_mystery / total_transactions

# Calculating expected probability
expected = P_fiction * P_mystery

# Displaying fiction probability
print(f"P(Fiction) = {P_fiction:.4f}")

# Displaying mystery probability
print(f"P(Mystery) = {P_mystery:.4f}")

# Displaying joint probability
print(f"P(Fiction ∩ Mystery) = {P_both:.4f}")

# Displaying blank line
print()

# Displaying expected probability
print(f"P(Fiction) × P(Mystery) = {expected:.4f}")

# Displaying blank line
print()

# Checking independence and correlation
if abs(P_both - expected) < 1e-10:
    print("Conclusion : Fiction and Mystery purchases are Independent.")
elif P_both > expected:
    print("Conclusion : Fiction and Mystery purchases are NOT Independent.")
    print("Correlation : Positive Correlation")
else:
    print("Conclusion : Fiction and Mystery purchases are NOT Independent.")
    print("Correlation : Negative Correlation")



(b) Independence and Correlation Analysis
P(Fiction) = 0.5000
P(Mystery) = 0.4800
P(Fiction ∩ Mystery) = 0.3600

P(Fiction) × P(Mystery) = 0.2400

Conclusion : Fiction and Mystery purchases are NOT Independent.
Correlation : Positive Correlation


**Interestingness Measures**

Interestingness was computed to determine the strength and quality of the fiction and mystery book purchase relationship. The values obtained allowed the different measures to be compared and to gain a better understanding of the relationship between the two categories.

In [8]:
# Displaying heading for interestingness measures
print("(c) Interestingness Measures")

# Calculating all confidence
all_confidence = fiction_and_mystery / max(fiction, mystery)

# Calculating confidence from fiction to mystery
confidence_F_M = fiction_and_mystery / fiction

# Calculating confidence from mystery to fiction
confidence_M_F = fiction_and_mystery / mystery

# Calculating maximum confidence
max_confidence = max(confidence_F_M, confidence_M_F)

# Calculating Kulczynski measure
kulczynski = (confidence_F_M + confidence_M_F) / 2

# Calculating cosine measure
cosine = fiction_and_mystery / math.sqrt(fiction * mystery)

# Calculating lift value
lift = confidence_F_M / (mystery / total_transactions)

# Storing contingency table values
A = fiction_and_mystery
B = not_fiction_and_mystery
C = fiction_and_not_mystery
D = not_fiction_and_not_mystery

# Calculating phi coefficient
phi = ((A * D) - (B * C)) / math.sqrt((A + B) * (C + D) * (A + C) * (B + D))

# Displaying all confidence
print(f"All Confidence : {all_confidence:.4f}")

# Displaying maximum confidence
print(f"Max Confidence : {max_confidence:.4f}")

# Displaying Kulczynski measure
print(f"Kulczynski     : {kulczynski:.4f}")

# Displaying cosine measure
print(f"Cosine         : {cosine:.4f}")

# Displaying lift value
print(f"Lift           : {lift:.4f}")

# Displaying phi coefficient
print(f"Phi Coefficient: {phi:.4f}")



(c) Interestingness Measures
All Confidence : 0.7200
Max Confidence : 0.7500
Kulczynski     : 0.7350
Cosine         : 0.7348
Lift           : 1.5000
Phi Coefficient: 0.4804


#### The calculated measures suggest that there is a significant positive correlation between the Fiction and Mystery books purchased. Association rule Fiction Mystery is a strong association rule with its support (36%) and confidence (72%) exceeding the minimum requirements. The lift value of 1.50 implies that customers who buy Fiction books are 1.5 times more likely to buy Mystery books than would happen by chance; a positive value of the phi coefficient (0.442) suggests a moderate positive association between the two classes of books. What's more, the high All Confidence (0.72), Max Confidence (0.75), Kulczynski (0.735) and Cosine (0.735) values also indicate a good and stable relationship. Based on these findings, bookstores can utilize this relationship to cross sell, product bundling, promotions targeting specific customers and better bookstore inventory management.

### 3: Decision Tree Construction

**Create Dataset**

The employee data was represented in a structured data table and the records were enriched according to their frequencies. A new data set size was then shown to ensure that all the repeated data sets had been shown for further analysis.

In [16]:
# Creating employee training dataset
data = {
    'Department':['Sales','Sales','Sales','IT','IT','IT','IT','HR','HR','Admin','Admin'],
    'Status':['Senior','Junior','Junior','Junior','Senior','Junior','Senior','Senior','Junior','Senior','Junior'],
    'Age':['31-35','26-30','31-35','21-25','36-40','26-30','41-45','36-40','31-35','46-50','26-30'],
    'Salary':['51-55','26-30','31-35','41-45','61-65','46-50','61-65','46-50','36-40','36-40','26-30'],
    'Count':[25,35,45,15,8,5,7,12,6,3,7]
}

# Creating dataframe
df = pd.DataFrame(data)

# Displaying training dataset
print(df)

# Expanding rows using count values
expanded_df = df.loc[df.index.repeat(df["Count"])].reset_index(drop=True)

# Removing count column
expanded_df = expanded_df.drop(columns=["Count"])

# Displaying total expanded records
print("Total Records after Expansion =", len(expanded_df))

   Department  Status    Age Salary  Count
0       Sales  Senior  31-35  51-55     25
1       Sales  Junior  26-30  26-30     35
2       Sales  Junior  31-35  31-35     45
3          IT  Junior  21-25  41-45     15
4          IT  Senior  36-40  61-65      8
5          IT  Junior  26-30  46-50      5
6          IT  Senior  41-45  61-65      7
7          HR  Senior  36-40  46-50     12
8          HR  Junior  31-35  36-40      6
9       Admin  Senior  46-50  36-40      3
10      Admin  Junior  26-30  26-30      7
Total Records after Expansion = 168


**Decision Tree Construction**

**Entropy Function**

The entropy value was computed to quantify the impurity of target values of the data set. This value was subsequently taken as a measure of the split value to be used in the construction of the tree.

In [17]:
# Calculating entropy value
def entropy(target):

    # Counting class probabilities
    values = target.value_counts(normalize=True)

    # Initializing entropy value
    ent = 0

    # Calculating entropy
    for p in values:
        ent -= p * log2(p)

    # Returning entropy value
    return ent

**Information Gain**

The information gain of each attribute was computed by subtracting the information gain of each split from the overall information gain. The resulting values were used to determine the most appropriate attribute to use to build the decision tree.

In [18]:
# Calculating information gain
def information_gain(data, feature, target):

    # Calculating total entropy
    total_entropy = entropy(data[target])

    # Getting unique feature values
    values = data[feature].unique()

    # Initializing weighted entropy
    weighted_entropy = 0

    # Calculating weighted entropy
    for value in values:

        # Selecting matching records
        subset = data[data[feature] == value]

        # Adding weighted entropy
        weighted_entropy += (len(subset) / len(data)) * entropy(subset[target])

    # Returning information gain
    return total_entropy - weighted_entropy

**Total Entropy**

The total entropy and information gain value for each attribute were computed to determine the best splitting attribute. The results were used to determine the attribute used for most information gain to build a decision tree.

In [20]:
# Calculating total entropy of status values
total_entropy = entropy(expanded_df["Status"])

# Displaying total entropy value
print(f"Total Entropy = {total_entropy:.4f}")

# Displaying information gain heading
print("\nInformation Gain")

# Defining features for calculation
features = ["Department","Age","Salary"]

# Calculating information gain for each feature
for feature in features:

    # Calculating feature information gain
    gain = information_gain(expanded_df, feature, "Status")

    # Displaying information gain value
    print(f"{feature:12s} : {gain:.4f}")

Total Entropy = 0.9122

Information Gain
Department   : 0.0612
Age          : 0.4988
Salary       : 0.7746


**Encode Data**

Label Encoding was used to transform categorical attributes to quantitative data to enable machine learning processing. Features from the input and target variable were separated to create the data for training the classification model.

In [21]:
# Creating label encoders for categorical columns
dept_encoder = LabelEncoder()
age_encoder = LabelEncoder()
salary_encoder = LabelEncoder()
status_encoder = LabelEncoder()

# Converting department values into numeric format
expanded_df["Department"] = dept_encoder.fit_transform(expanded_df["Department"])

# Converting age values into numeric format
expanded_df["Age"] = age_encoder.fit_transform(expanded_df["Age"])

# Converting salary values into numeric format
expanded_df["Salary"] = salary_encoder.fit_transform(expanded_df["Salary"])

# Converting status values into numeric format
expanded_df["Status"] = status_encoder.fit_transform(expanded_df["Status"])

# Selecting input features
X = expanded_df[["Department","Age","Salary"]]

# Selecting target class
y = expanded_df["Status"]

**Build Decision Tree**

The entropy-based splitting criterion was used to build the decision tree model and trained with the prepared dataset. The tree rules generated are shown for the decision making process with the chosen attributes.

In [22]:
# Creating decision tree classifier using entropy criterion
tree = DecisionTreeClassifier(
    criterion="entropy",
    random_state=0
)

# Training the decision tree model
tree.fit(X, y)

# Displaying decision tree rules heading
print("\nDecision Tree Rules\n")

# Extracting decision tree rules
tree_rules = export_text(
    tree,
    feature_names=["Department","Age","Salary"]
)

# Displaying decision tree rules
print(tree_rules)


Decision Tree Rules

|--- Salary <= 3.50
|   |--- Age <= 3.50
|   |   |--- class: 0
|   |--- Age >  3.50
|   |   |--- class: 1
|--- Salary >  3.50
|   |--- Age <= 1.50
|   |   |--- class: 0
|   |--- Age >  1.50
|   |   |--- class: 1



**Decision Node Selection**

The feature importance values were calculated to identify the contribution of each attribute in the decision tree model. The attribute that had the most important value was chosen to be the top node for the classification decision making.

In [23]:
# Displaying decision node selection heading
print("\nDecision Node Selection")

# Getting feature importance values
importance = tree.feature_importances_

# Displaying importance score for each feature
for f, score in zip(["Department","Age","Salary"], importance):

    # Printing feature importance
    print(f"{f:12s} : {score:.4f}")

# Displaying root node selection information
print("\nFeature with highest information gain becomes the Root Node.")


Decision Node Selection
Department   : 0.0000
Age          : 0.2893
Salary       : 0.7107

Feature with highest information gain becomes the Root Node.


**Naive Bayesian Classification**

The Naïve Bayes classification model was trained with the employee data that was prepared and predicted the status of a new data record. The class probabilities were computed and the final classification was determined by the largest probability value.

In [25]:
# Creating Naive Bayes classifier
nb = CategoricalNB()

# Training Naive Bayes model
nb.fit(X, y)

# Creating new data tuple for prediction
new_tuple = pd.DataFrame({
    "Department":["IT"],
    "Age":["26-30"],
    "Salary":["46-50"]
})

# Converting department value into numeric format
new_tuple["Department"] = dept_encoder.transform(new_tuple["Department"])

# Converting age value into numeric format
new_tuple["Age"] = age_encoder.transform(new_tuple["Age"])

# Converting salary value into numeric format
new_tuple["Salary"] = salary_encoder.transform(new_tuple["Salary"])

# Predicting class for new tuple
prediction = nb.predict(new_tuple)

# Calculating class probabilities
probabilities = nb.predict_proba(new_tuple)

# Converting predicted numeric class into original label
predicted_status = status_encoder.inverse_transform(prediction)

# Displaying new tuple details
print("New Tuple")

print("Department = IT")
print("Age        = 26-30")
print("Salary     = 46-50")

# Displaying class probability heading
print("\nClass Probabilities")

# Getting original class names
classes = status_encoder.inverse_transform([0,1])

# Displaying probability of each class
for cls, prob in zip(classes, probabilities[0]):

    print(f"{cls:8s} : {prob:.4f}")

# Displaying final predicted class
print("\nPredicted Class =", predicted_status[0])

# Displaying classification reason
print("\nReason")

print("The class having the highest posterior probability")
print("is selected as the final classification.")

New Tuple
Department = IT
Age        = 26-30
Salary     = 46-50

Class Probabilities
Junior   : 0.8886
Senior   : 0.1114

Predicted Class = Junior

Reason
The class having the highest posterior probability
is selected as the final classification.


#### The Naïve Bayes classifier predicts Junior since the posterior probability of it is greater than the posterior probability of Senior. The probability for the Senior class becomes zero since there are no Senior employees in the training data with the age group 26...30, whereas the combination of attributes exists for the Junior class. In this case, it is likely that the employee is in the Junior category.

### 4 K-Means Clustering with Euclidean Distance

The customer location points and initial cluster centers were determined to apply the K-Means clustering algorithm. The initial centers selected were ready to calculate distance and assign data points to clusters.




In [29]:
# Creating customer location points
points = {
    "P1": (3, 12),
    "P2": (3, 6),
    "P3": (9, 3),
    "Q1": (6, 9),
    "Q2": (8, 6),
    "Q3": (7, 3),
    "R1": (2, 3),
    "R2": (5, 11)
}

# Defining initial cluster centers
centers = {
    "C1": np.array([3,12]),   
    "C2": np.array([6,9]),   
    "C3": np.array([2,3])     
}

**Euclidean Distance Function**

The Euclidean distance calculation method was stipulated to measure the distance between the data points and the cluster centers. This distance calculation was used to determine the centroids in K-Means clustering.

In [30]:
# Calculating Euclidean distance between point and cluster center
def euclidean(point, center):

    # Returning distance value using Euclidean formula
    return np.sqrt(np.sum((point - center)**2))

**One Iteration of K-Means**

The Euclidean distance method was used to calculate the distance between the data points and the initial cluster centers. They were assigned to the closest cluster which corresponds to the minimum value of the distance, and the clustering results were shown in a table.

In [31]:
# Creating dictionary to store cluster assignments
assignments = {}

# Creating list to store distance details
rows = []

# Calculating distance for each data point
for name, point in points.items():

    # Converting point into array format
    point = np.array(point)

    # Calculating distance from each cluster center
    d1 = euclidean(point, centers["C1"])
    d2 = euclidean(point, centers["C2"])
    d3 = euclidean(point, centers["C3"])

    # Storing calculated distances
    distances = [d1, d2, d3]

    # Finding nearest cluster center
    cluster = np.argmin(distances) + 1

    # Storing point cluster assignment
    assignments[name] = cluster

    # Adding distance details to list
    rows.append([
        name,
        round(d1,2),
        round(d2,2),
        round(d3,2),
        "Cluster " + str(cluster)
    ])

# Creating dataframe for distance results
distance_table = pd.DataFrame(
    rows,
    columns=["Point",
             "Distance to C1",
             "Distance to C2",
             "Distance to C3",
             "Assigned Cluster"]
)

# Displaying distance table
print(distance_table)

  Point  Distance to C1  Distance to C2  Distance to C3 Assigned Cluster
0    P1            0.00            4.24            9.06        Cluster 1
1    P2            6.00            4.24            3.16        Cluster 3
2    P3           10.82            6.71            7.00        Cluster 2
3    Q1            4.24            0.00            7.21        Cluster 2
4    Q2            7.81            3.61            6.71        Cluster 2
5    Q3            9.85            6.08            5.00        Cluster 3
6    R1            9.06            7.21            0.00        Cluster 3
7    R2            2.24            2.24            8.54        Cluster 1


**Calculate New Cluster Centers**

The new cluster centers were computed as the average of the co-ordinates of all the points in each cluster. The new centers obtained following the initial iteration of K-Means were presented to further the K-Means clustering process.

In [32]:
# Displaying new cluster centers heading
print("\nNew Cluster Centers After First Iteration\n")

# Creating dictionary to store updated centers
new_centers = {}

# Calculating new center for each cluster
for cluster in [1,2,3]:

    # Creating list to store cluster points
    cluster_points = []

    # Collecting points assigned to current cluster
    for point_name, c in assignments.items():

        if c == cluster:
            cluster_points.append(points[point_name])

    # Converting cluster points into array format
    cluster_points = np.array(cluster_points)

    # Calculating mean value to find new center
    center = cluster_points.mean(axis=0)

    # Storing updated cluster center
    new_centers["C" + str(cluster)] = center

    # Displaying cluster details
    print("Cluster", cluster)

    print("Points :", cluster_points)

    print("New Center :", center)

    # Displaying blank line
    print()


New Cluster Centers After First Iteration

Cluster 1
Points : [[ 3 12]
 [ 5 11]]
New Center : [ 4.  11.5]

Cluster 2
Points : [[9 3]
 [6 9]
 [8 6]]
New Center : [7.66666667 6.        ]

Cluster 3
Points : [[3 6]
 [7 3]
 [2 3]]
New Center : [4. 4.]



**Continue Until Convergence**

The cluster assignments and the cluster centers were updated, and the clustering process was repeated until the values of the clusters centers were stable. The convergence of the algorithm was determined by the absence of any movement of the centers of the clusters from one iteration to the next.

In [33]:
# Updating centers with new cluster centers
centers = new_centers

# Starting iteration count
iteration = 2

# Running k means until convergence
while True:

    # Creating dictionary to store new assignments
    assignments = {}

    # Assigning points to nearest cluster center
    for name, point in points.items():

        # Converting point into array format
        point = np.array(point)

        # Creating list to store distances
        distances = []

        # Calculating distance from each center
        for center in centers.values():

            distances.append(euclidean(point, center))

        # Finding nearest cluster
        cluster = np.argmin(distances) + 1

        # Storing cluster assignment
        assignments[name] = cluster

    # Creating dictionary for updated centers
    updated_centers = {}

    # Calculating new centers for each cluster
    for cluster in [1,2,3]:

        # Creating list to store cluster points
        cluster_points = []

        # Collecting points belonging to cluster
        for point_name, c in assignments.items():

            if c == cluster:

                cluster_points.append(points[point_name])

        # Converting points into array format
        cluster_points = np.array(cluster_points)

        # Calculating updated cluster center
        updated_centers["C" + str(cluster)] = cluster_points.mean(axis=0)

    # Checking whether centers remain unchanged
    if all(np.allclose(updated_centers[key], centers[key])
           for key in centers):

        break

    # Updating centers for next iteration
    centers = updated_centers

    # Increasing iteration count
    iteration += 1

# Displaying convergence result
print("Algorithm Converged After", iteration, "Iterations\n")

Algorithm Converged After 3 Iterations



**Final Clusters**

A plot of the final clusters and corresponding centre points has been shown when the K-Means algorithm converged. Assigned data points and final cluster centers were displayed to illustrate the clustering results obtained.


In [34]:
# Displaying final clusters and centers
for cluster in [1,2,3]:

    # Displaying cluster number
    print("Cluster", cluster)

    # Creating list to store cluster members
    members = []

    # Finding points assigned to current cluster
    for point, c in assignments.items():

        if c == cluster:

            members.append(point)

    # Displaying points in cluster
    print("Points :", members)

    # Displaying final cluster center
    print("Final Center :", centers["C" + str(cluster)])

    # Displaying blank line
    print()

Cluster 1
Points : ['P1', 'Q1', 'R2']
Final Center : [ 4.66666667 10.66666667]

Cluster 2
Points : ['P3', 'Q2', 'Q3']
Final Center : [8. 4.]

Cluster 3
Points : ['P2', 'R1']
Final Center : [2.5 4.5]



**Final Summary Table**

The final cluster assignments were set up in a tabular format reflecting each data point and the cluster it belonged to. The brief presented a clear summary of the results obtained on K-Means clustering after convergence.


In [36]:
# Creating final cluster assignment dataframe
summary = pd.DataFrame({
    "Point": list(assignments.keys()),
    "Cluster": list(assignments.values())
})

# Displaying final cluster assignment heading
print("Final Cluster Assignment")

# Displaying cluster assignment table
print(summary)

Final Cluster Assignment
  Point  Cluster
0    P1        1
1    P2        3
2    P3        2
3    Q1        1
4    Q2        2
5    Q3        2
6    R1        3
7    R2        1
